`no-restricted-properties`: does not work with destructuring

**Tell us about your environment**

* **ESLint Version:**
v3.5.0
* **Node Version:**
v6.5.0
* **npm Version:**
3.something

**What parser (default, Babel-ESLint, etc.) are you using?**
default

**Please show your full configuration:**
n/a

**What did you do? Please include the actual source code causing the issue.**
```sh
echo 'const { bar } = foo; const bar = foo.bar;' | ./node_modules/.bin/eslint --stdin --parser-options '{"ecmaVersion": 6}' --rule 'no-restricted-properties:[2,{"object":"foo","property":"bar"}]' --rule no-undef:0 --rule no-unused-vars:0 --no-eslintrc
```

**What did you expect to happen?**
```sh
<text>
  1:##  error  'foo.bar' is restricted from being used  no-restricted-properties
  1:34  error  'foo.bar' is restricted from being used  no-restricted-properties

✖ 2 problems (2 errors, 0 warnings)
```

**What actually happened? Please include the actual, raw output from ESLint.**
```sh
<text>
  1:34  error  'foo.bar' is restricted from being used  no-restricted-properties

✖ 1 problem (1 error, 0 warnings)
```

Per https://github.com/eslint/eslint/pull/7139#issuecomment-246749114

--- a/lib/rules/no-restricted-properties.js
+++ b/lib/rules/no-restricted-properties.js
@@ -91,31 +91,73 @@ module.exports = {
             }
         });
 
-        return {
-            MemberExpression(node) {
-                const objectName = node.object && node.object.name;
-                const propertyName = astUtils.getStaticPropertyName(node);
-                const matchedObject = restrictedProperties.get(objectName);
-                const matchedObjectProperty = matchedObject ? matchedObject.get(propertyName) : globallyRestrictedObjects.get(objectName);
-                const globalMatchedProperty = globallyRestrictedProperties.get(propertyName);
+        /**
+        * Checks to see whether a property access is restricted, and reports it if so.
+        * @param {ASTNode} node The node to report
+        * @param {string} objectName The name of the object
+        * @param {string} propertyName The name of the property
+        * @returns {undefined}
+        */
+        function checkPropertyAccess(node, objectName, propertyName) {
+            if (propertyName === null) {
+                return;
+            }
+            const matchedObject = restrictedProperties.get(objectName);
+            const matchedObjectProperty = matchedObject ? matchedObject.get(propertyName) : globallyRestrictedObjects.get(objectName);
+            const globalMatchedProperty = globallyRestrictedProperties.get(propertyName);
 
-                if (matchedObjectProperty) {
-                    const message = matchedObjectProperty.message ? " " + matchedObjectProperty.message : "";
+            if (matchedObjectProperty) {
+                const message = matchedObjectProperty.message ? " " + matchedObjectProperty.message : "";
 
-                    context.report(node, "'{{objectName}}.{{propertyName}}' is restricted from being used.{{message}}", {
-                        objectName,
-                        propertyName,
-                        message
-                    });
-                } else if (globalMatchedProperty) {
-                    const message = globalMatchedProperty.message ? " " + globalMatchedProperty.message : "";
+                context.report(node, "'{{objectName}}.{{propertyName}}' is restricted from being used.{{message}}", {
+                    objectName,
+                    propertyName,
+                    message
+                });
+            } else if (globalMatchedProperty) {
+                const message = globalMatchedProperty.message ? " " + globalMatchedProperty.message : "";
 
-                    context.report(node, "'{{propertyName}}' is restricted from being used.{{message}}", {
-                        propertyName,
-                        message
+                context.report(node, "'{{propertyName}}' is restricted from being used.{{message}}", {
+                    propertyName,
+                    message
+                });
+            }
+        }
+
+        /**
+        * Checks property accesses in a destructuring assignment expression, e.g. `var foo; ({foo} = bar);`
+        * @param {ASTNode} node An AssignmentExpression or AssignmentPattern node
+        * @returns {undefined}
+        */
+        function checkDestructuringAssignment(node) {
+            if (node.right.type === "Identifier") {
+                const objectName = node.right.name;
+
+                if (node.left.type === "ObjectPattern") {
+                    node.left.properties.forEach(property => {
+                        checkPropertyAccess(node.left, objectName, astUtils.getStaticPropertyName(property));
                     });
                 }
             }
+        }
+
+        return {
+            MemberExpression(node) {
+                checkPropertyAccess(node, node.object && node.object.name, astUtils.getStaticPropertyName(node));
+            },
+            VariableDeclarator(node) {
+                if (node.init && node.init.type === "Identifier") {
+                    const objectName = node.init.name;
+
+                    if (node.id.type === "ObjectPattern") {
+                        node.id.properties.forEach(property => {
+                            checkPropertyAccess(node.id, objectName, astUtils.getStaticPropertyName(property));
+                        });
+                    }
+                }
+            },
+            AssignmentExpression: checkDestructuringAssignment,
+            AssignmentPattern: checkDestructuringAssignment
         };
     }
 };

--- a/tests/lib/rules/no-restricted-properties.js
+++ b/tests/lib/rules/no-restricted-properties.js
@@ -95,6 +95,74 @@ ruleTester.run("no-restricted-properties", rule, {
             options: [{
                 object: "foo"
             }]
+        }, {
+            code: "let bar = foo;",
+            options: [{ object: "foo", property: "bar" }],
+            parserOptions: { ecmaVersion: 6 },
+        }, {
+            code: "let {baz: bar} = foo;",
+            options: [{ object: "foo", property: "bar" }],
+            parserOptions: { ecmaVersion: 6 }
+        }, {
+            code: "let {unrelated} = foo;",
+            options: [{ object: "foo", property: "bar" }],
+            parserOptions: { ecmaVersion: 6 }
+        }, {
+            code: "let {baz: {bar: qux}} = foo;", property: "bar",
+            options: [{ object: "foo", property: "bar" }],
+            parserOptions: { ecmaVersion: 6 }
+        }, {
+            code: "let {bar} = foo.baz;",
+            options: [{ object: "foo", property: "bar" }],
+            parserOptions: { ecmaVersion: 6 }
+        }, {
+            code: "let {baz: bar} = foo;",
+            options: [{ property: "bar" }],
+            parserOptions: { ecmaVersion: 6 }
+        }, {
+            code: "let baz; ({baz: bar} = foo)",
+            options: [{ object: "foo", property: "bar" }],
+            parserOptions: { ecmaVersion: 6 }
+        }, {
+            code: "let bar;",
+            options: [{ object: "foo", property: "bar" }],
+            parserOptions: { ecmaVersion: 6 }
+        }, {
+            code: "let bar; ([bar = 5] = foo);",
+            options: [{ object: "foo", property: "1" }],
+            parserOptions: { ecmaVersion: 6 }
+        }, {
+            code: "function qux({baz: bar} = foo) {}",
+            options: [{ object: "foo", property: "bar" }],
+            parserOptions: { ecmaVersion: 6 }
+        }, {
+            code: "let [bar, baz] = foo;",
+            options: [{ object: "foo", property: "1" }],
+            parserOptions: { ecmaVersion: 6 }
+        }, {
+            code: "let [, bar] = foo;",
+            options: [{ object: "foo", property: "0" }],
+            parserOptions: { ecmaVersion: 6 }
+        }, {
+            code: "let [, bar = 5] = foo;",
+            options: [{ object: "foo", property: "1" }],
+            parserOptions: { ecmaVersion: 6 }
+        }, {
+            code: "let bar; ([bar = 5] = foo);",
+            options: [{ object: "foo", property: "0" }],
+            parserOptions: { ecmaVersion: 6 }
+        }, {
+            code: "function qux([bar] = foo) {}",
+            options: [{ object: "foo", property: "0" }],
+            parserOptions: { ecmaVersion: 6 }
+        }, {
+            code: "function qux([, bar] = foo) {}",
+            options: [{ object: "foo", property: "0" }],
+            parserOptions: { ecmaVersion: 6 }
+        }, {
+            code: "function qux([, bar] = foo) {}",
+            options: [{ object: "foo", property: "1" }],
+            parserOptions: { ecmaVersion: 6 }
         }
     ],
 
@@ -199,6 +267,66 @@ ruleTester.run("no-restricted-properties", rule, {
                 message: "'require.resolve' is restricted from being used.",
                 type: "MemberExpression"
             }]
+        }, {
+            code: "let {bar} = foo;",
+            options: [{ object: "foo", property: "bar" }],
+            parserOptions: { ecmaVersion: 6 },
+            errors: [{ message: "'foo.bar' is restricted from being used.", type: "ObjectPattern" }]
+        }, {
+            code: "let {bar: baz} = foo;",
+            options: [{ object: "foo", property: "bar" }],
+            parserOptions: { ecmaVersion: 6 },
+            errors: [{ message: "'foo.bar' is restricted from being used.", type: "ObjectPattern" }]
+        }, {
+            code: "let {'bar': baz} = foo;",
+            options: [{ object: "foo", property: "bar" }],
+            parserOptions: { ecmaVersion: 6 },
+            errors: [{ message: "'foo.bar' is restricted from being used.", type: "ObjectPattern" }]
+        }, {
+            code: "let {bar: {baz: qux}} = foo;",
+            options: [{ object: "foo", property: "bar" }],
+            parserOptions: { ecmaVersion: 6 },
+            errors: [{ message: "'foo.bar' is restricted from being used.", type: "ObjectPattern" }]
+        }, {
+            code: "let {bar} = foo;",
+            options: [{ object: "foo" }],
+            parserOptions: { ecmaVersion: 6 },
+            errors: [{ message: "'foo.bar' is restricted from being used.", type: "ObjectPattern" }]
+        }, {
+            code: "let {bar: baz} = foo;",
+            options: [{ object: "foo" }],
+            parserOptions: { ecmaVersion: 6 },
+            errors: [{ message: "'foo.bar' is restricted from being used.", type: "ObjectPattern" }]
+        }, {
+            code: "let {bar} = foo;",
+            options: [{ property: "bar" }],
+            parserOptions: { ecmaVersion: 6 },
+            errors: [{ message: "'bar' is restricted from being used.", type: "ObjectPattern" }]
+        }, {
+            code: "let bar; ({bar} = foo);",
+            options: [{ object: "foo", property: "bar" }],
+            parserOptions: { ecmaVersion: 6 },
+            errors: [{ message: "'foo.bar' is restricted from being used.", type: "ObjectPattern" }]
+        }, {
+            code: "let bar; ({bar: baz = 1} = foo);",
+            options: [{ object: "foo", property: "bar" }],
+            parserOptions: { ecmaVersion: 6 },
+            errors: [{ message: "'foo.bar' is restricted from being used.", type: "ObjectPattern" }]
+        }, {
+            code: "function qux({bar} = foo) {}",
+            options: [{ object: "foo", property: "bar" }],
+            parserOptions: { ecmaVersion: 6 },
+            errors: [{ message: "'foo.bar' is restricted from being used.", type: "ObjectPattern" }]
+        }, {
+            code: "function qux({bar: baz} = foo) {}",
+            options: [{ object: "foo", property: "bar" }],
+            parserOptions: { ecmaVersion: 6 },
+            errors: [{ message: "'foo.bar' is restricted from being used.", type: "ObjectPattern" }]
+        }, {
+            code: "var {['foo']: qux, bar} = baz",
+            options: [{ object: "baz", property: "foo" }],
+            parserOptions: { ecmaVersion: 6 },
+            errors: [{ message: "'baz.foo' is restricted from being used.", type: "ObjectPattern" }]
         }
     ]
 });

